eos-tech-support/eos-stage-ostree

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash -e
+
+function usage {
+    cat <<EOF
+Usage:
+    $0 STAGE [PASSWORD]
+Arguments:
+    STAGE     One of the following:
+                - dev: development staging (requires password)
+                       (note that the dev stage does not automatically
+                       update to the next major release series)
+                - demo: beta testing in advance of production releases
+                - prod: full production releases for general availability
+    PASSWORD  Development password (for staging branch only)
+EOF
+}
+
+if [ $# -lt 1 ] || [ $# -gt 2 ] ; then
+    if [ $# -lt 1 ] ; then
+        echo "Error: missing STAGE argument" >&2
+    else
+        echo "Error: extra arguments after STAGE PASSWORD" >&2
+    fi
+    echo >&2
+    usage >&2
+    exit 1
+fi
+
+STAGE="$1"
+
+PASSWORD=
+if [ $# -ge 2 ] ; then
+    PASSWORD="$2"
+fi
+
+if [ $EUID != 0 ] ; then
+    echo "Program requires superuser privileges" >&2
+    exit 1
+fi
+
+if [ "$STAGE" == "dev" ] ; then
+    if [ $# -lt 2 ] ; then
+	echo "Error: missing required PASSWORD for dev stage" >&2
+	exit 1
+    fi
+    base_url="https://endless:${PASSWORD}@origin.ostree.endlessm.com/staging/dev"
+    # Get the OS product.major version, as dev stages stay
+    # on the major version (e.g., eos3.2)
+    version=$(. /etc/os-release && echo $VERSION | cut -d. -f1-2)
+    # Check for an update every time the updater runs
+    interval_days=0
+elif [ "$STAGE" == "demo" ] ; then
+    base_url="https://ostree.endlessm.com/staging/demo"
+    # Get the OS product version (e.g., eos3)
+    version=$(. /etc/os-release && echo $VERSION | cut -d. -f1)
+    # Check for an update every time the updater runs
+    interval_days=0
+elif [ "$STAGE" == "prod" ] ; then
+    base_url="https://ostree.endlessm.com/ostree"
+    # Get the OS product version (e.g., eos3)
+    version=$(. /etc/os-release && echo $VERSION | cut -d. -f1)
+    # By default we check for updates every two weeks on production
+    interval_days=14
+else
+    echo "Invalid STAGE -- must be dev, demo, or prod" >&2
+    exit 1
+fi
+
+# Get the current refspec.
+refspec=$(ostree admin status | awk '/refspec:/{print $3}' | head -n1)
+branch=${refspec#*:}
+
+# Construct the new branch with the product or product.major version
+# depending on the stage.
+new_branch="${branch%/*}/eos${version}"
+
+# Get the current URL and convert to the new stage.
+url=$(ostree config get 'remote "eos".url')
+repo=${url##*/}
+new_url="${base_url}/${repo}"
+
+# Make sure the autoupdater doesn't start the updater after it's
+# killed below.
+echo "Stopping running eos-autoupdater."
+systemctl stop eos-autoupdater.timer eos-autoupdater.service
```

Comment:
The instructions I would provide would be to run this script and then manually check for updates from the settings detail page.  I'd prefer not to allow downgrades by default, as one use case is to download the latest eos3.2 staging image and configure it so that it will get the next beta release (but stick on the current staging in the meantime).  Also, if a user wants to switch back from demo to prod, I would not want to automatically downgrade the user immediately -- I'd prefer that the normal case is that they don't get any more beta updates, but they wait where they are for the next prod release.  If anyone really wants a downgrade, we can provide explicit instructions.

That said, if the user does not immediately check for the updates, I think it does make sense to re-enable the autoupdater timer.  And, I think it is fine to do that on dev, too.

So, unless you object, I'm just going to turn the autoupdater back on at the end, regardless of stage.